src/libraries/System.Security.Cryptography.Primitives/src/System/Security/Cryptography/CryptoStream.cs

```diff
@@ -630,12 +635,43 @@ private async ValueTask WriteAsyncCore(byte[] buffer, int offset, int count, Can
                     Debug.Assert(_inputBuffer != null);
                     // In this case, we don't have an entire block's worth left, so store it up in the
                     // input buffer, which by now must be empty.
-                    Buffer.BlockCopy(buffer, currentInputIndex, _inputBuffer, 0, bytesToWrite);
+                    buffer.Slice(currentInputIndex, bytesToWrite).CopyTo(_inputBuffer);
                     _inputBufferIndex += bytesToWrite;
                     return;
                 }
             }
             return;
+
+            static int TransformBlock(ICryptoTransform transform, ReadOnlyMemory<byte> inputBuffer, byte[] outputBuffer, int outputOffset)
+            {
+                if (MemoryMarshal.TryGetArray(inputBuffer, out ArraySegment<byte> segment))
+                {
+                    // Skip the copy if readonlymemory is actually an array.
+                    Debug.Assert(segment.Array is not null);
+                    return transform.TransformBlock(segment.Array, segment.Offset, inputBuffer.Length, outputBuffer, outputOffset);
+                }
+                else
+                {
+                    var rentedBuffer = ArrayPool<byte>.Shared.Rent(inputBuffer.Length);
+                    try
+                    {
+                        inputBuffer.CopyTo(rentedBuffer);
+                        int result = transform.TransformBlock(rentedBuffer, 0, inputBuffer.Length, outputBuffer, outputOffset);
+                        CryptographicOperations.ZeroMemory(rentedBuffer.AsSpan(0, inputBuffer.Length));
+                        ArrayPool<byte>.Shared.Return(rentedBuffer);
+                        rentedBuffer = null;
+
+                        return result;
+                    }
+                    catch
```

Comment:
The catch block here is unnecessary. Instead, consider restructuring the outer try block as the following pseudocode:

```cs
byte[] rented = ArrayPool.Rent();
try
{
    input.CopyTo(rented);
    Transform(from: rented, to: output);
}
finally
{
    ZeroMem(rented);
}
ArrayPool.Return(rented);
```

By putting a finally block around the code where the rented buffer contains potentially sensitive data, we can ensure that the whole thing is zeroed out whether the operation completes successfully or fails.

We could also consider pinning the temporary buffer as part of this operation to provide further protection against copies of the data being made.

Note to other reviewers: Should we skip the array pool entirely and instead use pre-pinned arrays? If an adversary can force the crypto transform to fail, they can force the application to abandon a bunch of Gen2 arrays, which could lead to perf degradation.